### PR TITLE
Map: fix closure serialization bug

### PR DESCRIPTION
**Pull request by davidrans**
- Add failing test
- Map: fix hashing arrays with non-serializable components

qa_req 0

CC @sctice-ifixit

### CODE REVIEW

**Comment by davidrans** [comment]
Hmm, that test passed locally (after the fix), but it's not passing in CI...

File: src/Map.php [hunk 1]
```diff
@@ -241,7 +241,12 @@ private function getHash($key)
             // Take a hash of the serialized value to avoid PHP having to do a
             // long string comparison on each lookup (we already check the
             // actual array values for equality ourselves).
-            return md5(serialize($key));
+            try {
+                return md5(serialize($key));
+            } catch (\Exception $e) {
+                // Sometimes serialization fails because the array contains anonymous classes or closures.
+                return md5(serialize(array_map(fn($val) => $this->getHash($val), $key)));
```
Comment (danielbeardsley) [inline src/Map.php]: Wow. I had no idea it was doing this when you use an array as a key. Ugh. Maybe we can go back to using the native extension?

Comment (sctice-ifixit) [reply src/Map.php]: https://github.com/php-ds/ext-ds/blob/d42750d69beb684ec3ef7b3da48fba769ce57ffc/src/ds/ds_htable.c#L219-L239

Comment (danielbeardsley) [reply src/Map.php]: Oh. :-(

The message is: don't use large arrays as keys.

**Comment by sctice-ifixit** [comment]
> Hmm, that test passed locally (after the fix), but it's not passing in CI...

My first guess is different configs that result in an error being triggered instead of an exception, resulting in the `catch` not being activated, but it sure looks like it's failing due to an uncaught exception. Maybe it's being converted to an exception later by PHPUnit? Or something else?

You could try a version that _always_ hashes the array values to see whether that doesn't work as expected.

Separately, we might want to test a closure in an array in an array. I expect that to work, but it's easy to make mistakes here.

**Comment by davidrans** [comment]
dev_block :hand: still only working locally, not in GH actions :shrug:

**Comment by sctice-ifixit** [comment]
> still only working locally, not in GH actions 🤷

Hm. Looking at the implementation on GH, I'd expect using a closure as part of the key to fail when using the native extension. I wonder whether the action is running using the native extension? I don't see anything in the workflow that would disable it. That seems a little silly, but it _would_ explain the behavior we're seeing.

**Comment by davidrans** [comment]
Ha! Looks like you were right un_dev_block :+1:

**Comment by davidrans** [comment]
Now that it's passing would you prefer if I reverted fc9a22a @sctice-ifixit?

**Comment by sctice-ifixit** [comment]
> Now that it's passing would you prefer if I reverted [fc9a22a](https://github.com/iFixit/php-ds-polyfill/commit/fc9a22a8f2497eec40df6a0f1405ebdf1ebd630f) @sctice-ifixit?

Ugh, I don't know. I think the case where we need to recursively hash is rare, so it seems dumb to me to do it proactively, but maybe setting up the exception handler is _worse_ than 
recursively hashing :shrug: Or maybe recursively hashing is fast enough that it truly doesn't matter.

I guess recursively hashing is the correct and simple thing to do, so we should do that until there's a demonstrated issue.

File: src/Map.php [hunk 2]
```diff
@@ -241,7 +241,9 @@ private function getHash($key)
             // Take a hash of the serialized value to avoid PHP having to do a
             // long string comparison on each lookup (we already check the
             // actual array values for equality ourselves).
-            return md5(serialize($key));
+
+            // Sometimes serializing the whole array fails because the array contains anonymous classes or closures.
+            return md5(serialize(array_map(fn($val) => $this->getHash($val), $key)));
```
Comment (sctice-ifixit) [inline src/Map.php:246]: There's a fair amount going on here. I might separate the recursive hash step from the serialize + md5 step.

**Review by sctice-ifixit — APPROVED** [review]
CR :flags: I made a small suggestion, but you can take or leave it.

**Pull request merged**